File: react-native/src/models/Style.ts
import { Dimension } from './Layout';
import {
  FontStyle,
  FontWeight,
  TextDecoration,
  TextDimensionUnit,
  TextOverflow,
} from './enums';
import type { Background } from './Background';

export interface TextDimension {
  value: number;
  unit: TextDimensionUnit;
}

export interface TextShadow {
  color: string;
  offsetX: number;
  offsetY: number;
  blur: number;
}

export interface TextGradient {
  type: string;
  colors: string[];
  angle: number;
  stops?: number[];
}

export interface Style {
  // Text properties (cascade to children)
  textColor?: string;
  fontSize?: TextDimension;
  fontFamily?: string;
  fontWeight?: FontWeight;
  fontStyle?: FontStyle;
  lineHeight?: TextDimension;
  letterSpacing?: number;
  textDecoration?: TextDecoration;
  textAlign?: string;
  maxLines?: number;
  overflow?: TextOverflow;
  textShadow?: TextShadow;
  textGradient?: TextGradient;

  // Visual properties (do NOT cascade)
  background?: Background;
  backgroundColor?: string;

  // Border properties (do NOT cascade)
  borderRadius?: Dimension;
  borderWidth?: number;
  borderColor?: string;

  // Shadow properties (do NOT cascade)
  shadowColor?: string;
  shadowRadius?: number;
  shadowOffsetX?: number;
  shadowOffsetY?: number;

  // Universal (cascades)
  opacity?: number;
}

export interface StyleClass {
  name: string;
  style: Style;
}

export interface Theme {
  id: string;
  defaultStyle: Style;
  colors: Record<string, string>;
}

export const CASCADING_KEYS: (keyof Style)[] = [
  'textColor',
  'fontSize',
  'fontFamily',
  'fontWeight',
  'fontStyle',
  'lineHeight',
  'letterSpacing',
  'textDecoration',
  'textAlign',
  'maxLines',
  'overflow',
  'textShadow',
  'textGradient',
  'opacity',
];

export function mergeStyles(own: Style, fallback: Style): Style {
  const result: Style = { ...fallback };
  const ownKeys = Object.keys(own) as (keyof Style)[];
  for (const key of ownKeys) {
    if (own[key] !== undefined) {
      (result as Record<string, unknown>)[key] = own[key];
    }
  }
  return result;
}

export function cascadingOnly(style: Style): Style {
  const result: Partial<Style> = {};
  for (const key of CASCADING_KEYS) {
    if (style[key] !== undefined) {
      (result as Record<string, unknown>)[key] = style[key];
    }
  }
  return result as Style;
}

export function parseTextDimension(raw: unknown): TextDimension | undefined {
  if (raw == null) return undefined;
  if (typeof raw === 'number') return { value: raw, unit: 'platform' };
  if (typeof raw === 'object' && raw !== null) {
    const obj = raw as Record<string, unknown>;
    const value = typeof obj.value === 'number' ? obj.value : parseFloat(String(obj.value ?? 0));
    const unit: TextDimensionUnit = obj.unit === 'percent' ? 'percent' : 'platform';
    return { value: isNaN(value) ? 0 : value, unit };
  }
  return undefined;
}

export function resolveTextDimension(td: TextDimension, rootHeightPx: number): number {
  if (td.unit === 'percent') return rootHeightPx * td.value / 1000;
  return td.value;
}

File: flutter/lib/src/models/style.dart
import 'enums.dart';
import 'background.dart';
import 'layout.dart';
import 'text_dimension.dart';

class TextShadow {
  final String color;
  final double offsetX;
  final double offsetY;
  final double blur;

  const TextShadow({
    required this.color,
    this.offsetX = 0,
    this.offsetY = 0,
    this.blur = 0,
  });

  factory TextShadow.fromJson(Map<String, dynamic> json) => TextShadow(
        color: json['color'] as String? ?? '#00000040',
        offsetX: (json['offsetX'] as num?)?.toDouble() ?? 0,
        offsetY: (json['offsetY'] as num?)?.toDouble() ?? 0,
        blur: (json['blur'] as num?)?.toDouble() ?? 0,
      );
}

class TextGradient {
  final String type;
  final List<String> colors;
  final double angle;
  final List<double>? stops;

  const TextGradient({
    this.type = 'linear',
    required this.colors,
    this.angle = 0,
    this.stops,
  });

  factory TextGradient.fromJson(Map<String, dynamic> json) => TextGradient(
        type: json['type'] as String? ?? 'linear',
        colors: (json['colors'] as List?)?.map((e) => e as String).toList() ?? const [],
        angle: (json['angle'] as num?)?.toDouble() ?? 0,
        stops: (json['stops'] as List?)?.map((e) => (e as num).toDouble()).toList(),
      );
}

class Style {
  // Text properties (cascade to children)
  final String? textColor;
  final TextDimension? fontSize;
  final String? fontFamily;
  final NDFontWeight? fontWeight;
  final NDFontStyle? fontStyle;
  final TextDimension? lineHeight;
  final double? letterSpacing;
  final NDTextDecoration? textDecoration;
  final String? textAlign;
  final int? maxLines;
  final NDTextOverflow? overflow;
  final TextShadow? textShadow;
  final TextGradient? textGradient;

  // Visual properties (do NOT cascade)
  final Background? background;
  final String? backgroundColor;

  // Border properties (do NOT cascade)
  final Dimension? borderRadius;
  final double? borderWidth;
  final String? borderColor;

  // Shadow properties (do NOT cascade)
  final String? shadowColor;
  final double? shadowRadius;
  final double? shadowOffsetX;
  final double? shadowOffsetY;

  // Universal (cascades to children)
  final double? opacity;

  const Style({
    this.textColor,
    this.fontSize,
    this.fontFamily,
    this.fontWeight,
    this.fontStyle,
    this.lineHeight,
    this.letterSpacing,
    this.textDecoration,
    this.textAlign,
    this.maxLines,
    this.overflow,
    this.textShadow,
    this.textGradient,
    this.background,
    this.backgroundColor,
    this.borderRadius,
    this.borderWidth,
    this.borderColor,
    this.shadowColor,
    this.shadowRadius,
    this.shadowOffsetX,
    this.shadowOffsetY,
    this.opacity,
  });

  static const empty = Style();

  // this takes priority; fallback to other per field
  Style merge(Style? other) {
    if (other == null) return this;
    return Style(
      textColor: textColor ?? other.textColor,
      fontSize: fontSize ?? other.fontSize,
      fontFamily: fontFamily ?? other.fontFamily,
      fontWeight: fontWeight ?? other.fontWeight,
      fontStyle: fontStyle ?? other.fontStyle,
      lineHeight: lineHeight ?? other.lineHeight,
      letterSpacing: letterSpacing ?? other.letterSpacing,
      textDecoration: textDecoration ?? other.textDecoration,
      textAlign: textAlign ?? other.textAlign,
      maxLines: maxLines ?? other.maxLines,
      overflow: overflow ?? other.overflow,
      textShadow: textShadow ?? other.textShadow,
      textGradient: textGradient ?? other.textGradient,
      background: background ?? other.background,
      backgroundColor: backgroundColor ?? other.backgroundColor,
      borderRadius: borderRadius ?? other.borderRadius,
      borderWidth: borderWidth ?? other.borderWidth,
      borderColor: borderColor ?? other.borderColor,
      shadowColor: shadowColor ?? other.shadowColor,
      shadowRadius: shadowRadius ?? other.shadowRadius,
      shadowOffsetX: shadowOffsetX ?? other.shadowOffsetX,
      shadowOffsetY: shadowOffsetY ?? other.shadowOffsetY,
      opacity: opacity ?? other.opacity,
    );
  }

  // Returns only text properties + opacity (for cascading to children)
  Style cascadingOnly() => Style(
        textColor: textColor,
        fontSize: fontSize,
        fontFamily: fontFamily,
        fontWeight: fontWeight,
        fontStyle: fontStyle,
        lineHeight: lineHeight,
        letterSpacing: letterSpacing,
        textDecoration: textDecoration,
        textAlign: textAlign,
        maxLines: maxLines,
        overflow: overflow,
        textShadow: textShadow,
        textGradient: textGradient,
        opacity: opacity,
      );

  factory Style.fromJson(Map<String, dynamic> json) => Style(
        textColor: json['textColor'] as String?,
        fontSize: json['fontSize'] != null ? TextDimension.fromJson(json['fontSize']) : null,
        fontFamily: json['fontFamily'] as String?,
        fontWeight: json['fontWeight'] != null
            ? NDFontWeight.fromJson(json['fontWeight'] as String)
            : null,
        fontStyle: json['fontStyle'] != null
            ? NDFontStyle.fromJson(json['fontStyle'] as String)
            : null,
        lineHeight: json['lineHeight'] != null ? TextDimension.fromJson(json['lineHeight']) : null,
        letterSpacing: (json['letterSpacing'] as num?)?.toDouble(),
        textDecoration: json['textDecoration'] != null
            ? NDTextDecoration.fromJson(json['textDecoration'] as String)
            : null,
        textAlign: json['textAlign'] as String?,
        maxLines: json['maxLines'] as int?,
        overflow: json['overflow'] != null
            ? NDTextOverflow.fromJson(json['overflow'] as String)
            : null,
        textShadow: json['textShadow'] != null
            ? TextShadow.fromJson(json['textShadow'] as Map<String, dynamic>)
            : null,
        textGradient: json['textGradient'] != null
            ? TextGradient.fromJson(json['textGradient'] as Map<String, dynamic>)
            : null,
        background: json['background'] != null
            ? Background.fromJson(json['background'] as Map<String, dynamic>)
            : null,
        backgroundColor: json['backgroundColor'] as String?,
        borderRadius: json['borderRadius'] != null
            ? Dimension.fromJsonFlexible(json['borderRadius'])
            : null,
        borderWidth: (json['borderWidth'] as num?)?.toDouble(),
        borderColor: json['borderColor'] as String?,
        shadowColor: json['shadowColor'] as String?,
        shadowRadius: (json['shadowRadius'] as num?)?.toDouble(),
        shadowOffsetX: (json['shadowOffsetX'] as num?)?.toDouble(),
        shadowOffsetY: (json['shadowOffsetY'] as num?)?.toDouble(),
        opacity: (json['opacity'] as num?)?.toDouble(),
      );
}
